drivers/watchdog/wdt_wwdg_stm32.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2019 Centaur Analytics, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <watchdog.h>
+#include <soc.h>
+#include <errno.h>
+#include <assert.h>
+
+#include "wdt_wwdg_stm32.h"
+
+#define WWDG_INTERNAL_DIVIDER   4096U
+#define WWDG_RESET_LIMIT    WWDG_COUNTER_MIN
+#define WWDG_COUNTER_MIN    0x40
+#define WWDG_COUNTER_MAX    0x7f
+
+/* The timeout of the WWDG in milliseconds is calculated by the below formula:
+ *
+ * t_WWDG = 1000 * ((counter & 0x3F) + 1) / f_WWDG (ms)
+ *
+ * where:
+ *  - t_WWDG: WWDG timeout
+ *  - counter: a value in [0x40, 0x7F] representing the cycles before timeout.
+ *             Giving the counter a value below 0x40, will result in an
+ *             immediate system reset. A reset is produced when the counter
+ *             rolls over from 0x40 to 0x3F.
+ *  - f_WWDG: the frequency of the WWDG clock. This can be calculated by the
+ *            below formula:
+ *    f_WWDG = f_PCLK / (4096 * prescaler) (Hz)
+ *    where:
+ *     - f_PCLK: the clock frequency of the system
+ *     - 4096: the constant internal divider
+ *     - prescaler: the programmable divider with valid values of 1, 2, 4 or 8
+ *
+ * The minimum timeout is calculated with:
+ *  - counter = 0x40
+ *  - prescaler = 1
+ * The maximim timeout is calculated with:
+ *  - counter = 0x7F
+ *  - prescaler = 8
+ *
+ * E.g. for f_PCLK = 2MHz
+ *  t_WWDG_min = 1000 * ((0x40 & 0x3F) + 1) / (2000000 / (4096 * 1))
+ *             = 2.048 ms
+ *  t_WWDG_max = 1000 * ((0x7F & 0x3F) + 1) / (2000000 / (4096 * 8))
+ *             = 1048.576 ms
+ */
+
+#define ABS_DIFF_UINT(a, b)  ((a) > (b) ? (a) - (b) : (b) - (a))
+#define WWDG_TIMEOUT_ERROR_MARGIN   (100 * USEC_PER_MSEC)
+#define IS_WWDG_TIMEOUT(__TIMEOUT_GOLDEN__, __TIMEOUT__)  \
+	(ABS_DIFF_UINT(__TIMEOUT_GOLDEN__, __TIMEOUT__) < \
+	 WWDG_TIMEOUT_ERROR_MARGIN)
+
+static void wwdg_stm32_irq_config(struct device *dev);
+
+/**
+ * @brief Calculates the timeout in microseconds.
+ *
+ * @param prescaler The prescaler value.
+ * @param counter The counter value.
+ * @return The timeout calculated in microseconds.
+ */
+static u32_t wwdg_stm32_get_timeout(u32_t prescaler, u32_t counter)
+{
+	u32_t divider = WWDG_INTERNAL_DIVIDER * (1 << (prescaler >> 7));
+	float f_wwdg = SystemCoreClock / divider;
+
+	return USEC_PER_SEC * (((counter & 0x3F) + 1) / f_wwdg);
+}
+
+/**
+ * @brief Calculates prescaler & counter values.
+ *
+ * @param timeout Timeout value in microseconds.
+ * @param prescaler Pointer to prescaler value.
+ * @param counter Pointer to counter value.
+ */
+static void wwdg_stm32_convert_timeout(u32_t timeout,
+				       u32_t *prescaler,
+				       u32_t *counter)
+{
+	u32_t clock_freq = SystemCoreClock;
```

Comment:
Shouldn't we rather consider the APB clock (PCLK1) here?
Should differ from `SystemCoreClock` by AHB and APB1 prescalers.